### PR TITLE
CONSOLE-3740: Upgrade Cypress

### PR DESCRIPTION
**Pull request by rhamilto**
Based on my findings, there are a number of recommendations/observations:

As part of the upgrade from 8.5.0 >= 10.x, the existing cypress.json configs and test names need to be updated. Running `cypress open` with a >= 10.x version of Cypress installed provides a wizard that covers this process. I have done already done so for `packages/integration-tests-cypress` and `packages/operator-lifecycle-manager/integration-tests-cypress` in this PR. But the following configs and tests need updating:

* packages/topology/integration-tests/cypress.json
* packages/dev-console/integration-tests/cypress.json
* packages/helm-plugin/integration-tests/cypress.json
* packages/gitops-plugin/integration-tests /cypress.json
* packages/knative-plugin/integration-tests/cypress.json
* packages/pipelines-plugin/integration-tests /cypress.json
* packages/shipwright-plugin/integration-tests/cypress.json
* packages/webterminal-plugin/integration-tests/cypress.json
* ~~packages/ceph-storage-plugin/integration-tests-cypress/cypress-ceph.json~~ (will be removed per @SanjalKatiyar)

The tests run significantly faster in Cypress >= 10.x. This exposes an issue with tests that use `cy.visit` where the console redirects to `/` if the console isn't fully loaded before Cypress starts testing. A workaround is to visit `/` first and then check the perspective is set to `Administrator`. This appears to give the console time to fully load. I've made these adjustments for `packages/integration-tests-cypress` and `packages/operator-lifecycle-manager/integration-tests-cypress` via global command ([`cy.initAdmin`](https://github.com/openshift/console/pull/13070/files#diff-e3755d91635165a330d0287dbf943c3b40271114628b817946467673e2cd8d7d)) I added.

I've also added `cy.createProjectWithCLI` and `cy.deleteProjectWithCLI`. It occurs to me we should use the CLI to create and delete projects where testing these functions isn't important to the test in order to reduce the run time of the tests, which is a task from https://issues.redhat.com/browse/CONSOLE-3718.  cc: @TheRealJon 

### Notes on updated tests

Requires demo plugin be running locally:
* app/demo-dynamic-plugin.cy.ts

Require monitoring plugin be running locally:
* crud/other-routes.cy.ts
* i18n/pseudolocalization.cy.ts
* monitoring/monitoring.cy.ts

### CODE REVIEW

File: frontend/packages/integration-tests-cypress/tsconfig.json
Comment (TheRealJon) [inline frontend/packages/integration-tests-cypress/tsconfig.json:8]: This doesn't seem quite right. Should this be like the recommendation found here: https://docs.cypress.io/guides/tooling/typescript-support#Configure-tsconfigjson

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts:22]: What changed here?

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts:51]: Could we keep the name filtering and category filtering as separate test cases?
```ts
it('filters Operators by name', () {
  // Test name filtering
}
it('filters Operators by category, () {
  // Test category filtering
}
```

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.view.ts
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.view.ts:208]: We shouldn't need this: https://docs.cypress.io/api/commands/type#When-element-is-not-in-focus

Is there a bug here?

File: frontend/tsconfig.json
Comment (TheRealJon) [inline frontend/tsconfig.json:24]: I suppose this didn't work?
```suggestion
    "**/cypress.config.ts"
    "cypress",
  ],
```

**Review by TheRealJon — COMMENTED** [review]
Just a few observations. I don't fully understand why the ts config for the integration test package is weird. I remember trying to tinker with it at one point because VSCode kept telling me that 'cy' wasn't defined in the global scope, but I didn't have much luck with it. Let me know if you want to look at it together.

File: README.md
Comment (rhamilto) [reply README.md:231]: Good catch.  I initially had done so for CI when I thought `OAUTH_BASE_ADDRESS` differed there, too.  In trying to get CI working, I removed it entirely rather than moving it.

File: frontend/packages/integration-tests-cypress/cypress.config.js
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/cypress.config.js:25]: We certainly can, but given the scope of this task, I'm opting for "later".

File: frontend/packages/integration-tests-cypress/support/login.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/support/login.ts:22]: Good catch.  No, we don't.  Another leftover from when I thought CI needed a different `OAUTH_BASE_ADDRESS`, too.  I'll remove.

File: frontend/packages/integration-tests-cypress/support/selectors.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/support/selectors.ts:14]: Let me investigate these type changes.  When I was upgrading to 11.x and trying to get tests running, I changed these and didn't revisit.

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts
Comment (rhamilto) [reply frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts:22]: The new method of creating a project with the command line means the newly created project is not selected in the project picker (the existing test depended on project creation method).

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts
Comment (rhamilto) [reply frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts:51]: We can, but we'll have to refactor the test further since the existing tests have interdependence.

File: frontend/packages/integration-tests-cypress/support/login.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/support/login.ts:47]: I opted not to remove it because nearly every existing test is using it (in all packages).  But we certainly can.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.cy.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.cy.ts:71]: This was overlooked in https://github.com/openshift/console/pull/12666 an necessary to run the test locally.

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts
Comment (TheRealJon) [reply frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-hub.cy.ts:51]: Yeah, imo, that would be better than having one monolithic test, but I'll defer to you if it's worth the effort or not.

File: frontend/packages/integration-tests-cypress/cypress.config.js
Comment (TheRealJon) [reply frontend/packages/integration-tests-cypress/cypress.config.js:25]: Yeah, we can open a tech-debt story for this. Could you do that and add a link to the story in the comments here and in the other places this comment was generated?

File: frontend/packages/integration-tests-cypress/tsconfig.json
Comment (TheRealJon) [reply frontend/packages/integration-tests-cypress/tsconfig.json:8]: Yeah, I remember tinkering with this a while back and it was very temperamental. My gut says that we shouldn't need to directly include Cypress in our build like this. But if it works, we can revisit it later.

File: frontend/packages/integration-tests-cypress/support/login.ts
Comment (TheRealJon) [reply frontend/packages/integration-tests-cypress/support/login.ts:47]: I worry that someone unfamiliar with these changes or how sessions work would expect a call to `cy.logout()` log out of the console, but it would just be a no-op. I think it would be worth keeping the logout command, but it should actually log the current user out when used. It should be safe to remove its usage in most places now that we're using sessions.

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.view.ts
Comment (rhamilto) [reply frontend/packages/operator-lifecycle-manager/integration-tests-cypress/views/operator.view.ts:208]: The `input` is wrapped in `div` and doesn't get focus otherwise.  🤷

File: frontend/packages/integration-tests-cypress/support/login.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/support/login.ts:47]: Removed and added a standalone logout test.

File: frontend/packages/integration-tests-cypress/support/selectors.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/support/selectors.ts]: Updated to the correct type `<JQuery<HTMLElement>>`.

**Comment by rhamilto** [comment]
@TheRealJon, thanks for the review.  I believe I addressed all your initial comments.

File: frontend/packages/integration-tests-cypress/tests/dashboards/project-dashboard.cy.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/dashboards/project-dashboard.cy.ts:29]: Ah.  In CI the value is `system:admin` and not `kube:admin`, so this change is necessary.

Screenshot 2023-10-04 at 7 06 21 PM

File: frontend/packages/integration-tests-cypress/cypress.config.js
Comment (jpuzz0) [inline frontend/packages/integration-tests-cypress/cypress.config.js:27]: FWIW, although `.cy` is the recommended file extension for cypress files with this later version, to minimize the amount of changes for the migration initially, these config files could point to `.spec` extensions instead (aka, keep them as-is), and the file names could be updated as a separate change later to meet the latest cypress standards.

File: test-cypress.sh
Comment (jpuzz0) [inline test-cypress.sh:72]: Curious why these tests are being disabled, as I would assume these passing is the truest indicator that the upgrade to cypress v12 is a successful one.

Comment (rhamilto) [reply test-cypress.sh:72]: Because only on the `console` and `olm` tests have been updated.  See the description at the top of the PR.

File: frontend/packages/integration-tests-cypress/cypress.config.js
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/cypress.config.js:27]: That a good point, but I'd rather do it now since that change came with Cypress 10 and have it dealt with.  The renaming is done automatically by Cypress on first launch with the 10+ version.  We'd have to manually go back and do it later.

File: test-cypress.sh
Comment (jpuzz0) [reply test-cypress.sh:72]: Got it. Will these also be disabled as a part of the patternfly v5 upgrade? Or will these need to be fixed prior that change?

Comment (rhamilto) [reply test-cypress.sh:72]: Ideally before then.  I've started a discussion about the timing on Slack.  I will tag you there.

**Comment by rhamilto** [comment]
These changes have no user-facing impact, but assigning approvers as an FYI.

QE Approver: 
/assign @yanpzhan
Docs Approver: 
/assign @opayne1
PX Approver:
/assign @RickJWagner

**Review by RickJWagner — COMMENTED** [review]
Looks good.  Adding px-approved label.

File: frontend/packages/integration-tests-cypress/support/login.ts
Comment (TheRealJon) [inline frontend/packages/integration-tests-cypress/support/login.ts:55]: Could we pull this out into a view?

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts:36]: Do we need to do this before merging?

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts:91]: Is this test broken?

**Review by TheRealJon — COMMENTED** [review]
Just a few comments. Let me know your thoughts.

File: frontend/.eslintignore
Comment (rhamilto) [reply frontend/.eslintignore:12]: No.  There were linter errors that I didn't figure were worth resolving since this code is going away.

File: frontend/packages/integration-tests-cypress/tests/app/auth-multiuser-login.cy.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/auth-multiuser-login.cy.ts:17]: Needed after each test since they test different logins.

File: frontend/packages/integration-tests-cypress/tests/app/masthead.cy.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/masthead.cy.ts:55]: Probably, but this isn't functionally different than what it was doing before, so no change in behavior.  Follow on?

File: frontend/packages/integration-tests-cypress/support/login.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/support/login.ts:55]: No because of the limitation noted in the comment on line 43.  It's possible there is a way, but I couldn't figure out what it was...

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts
Comment (rhamilto) [reply frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts:36]: https://github.com/openshift/console/pull/13214 is ready to go to fix this.

File: frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts
Comment (rhamilto) [reply frontend/packages/operator-lifecycle-manager/integration-tests-cypress/tests/operator-uninstall.cy.ts:91]: Will fix as a follow on in https://github.com/openshift/console/pull/13214 (which will require a couple tweaks).  The PR was authored to work with the existing version of Cypress but didn't actually work because part of the fix comes with the Cypress upgrade.

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.cy.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.cy.ts:99]: Good catch.  I'll need to fix that.  Will be tomorrow.

File: frontend/packages/integration-tests-cypress/tests/app/masthead.cy.ts
Comment (TheRealJon) [reply frontend/packages/integration-tests-cypress/tests/app/masthead.cy.ts:55]: Yeah, I think that's fine. I just worry that with sessions, we won't know if the logout failed.

**Review by TheRealJon — APPROVED** [review]
/lgtm

We can fix the logout test as a follow-on.

**Comment by rhamilto** [comment]
Adding qe-approval label since this isn't customer facing.

/label qe-approval

**Comment by rhamilto** [comment]
/label qe-approved

typing is hard

File: frontend/packages/integration-tests-cypress/tests/app/masthead.cy.ts
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/masthead.cy.ts:55]: So after playing with this, I don't think there is a way to actually test logout fully because of the way `cy.session` works since it's interecepting the credentials.  Will leave it for now.

**Pull request merged**
